### ISSUE TITLE
Add an option to output the version that would be published

### ISSUE THREAD

**Issue by BenoitAverty** [semantic-release/semantic-release#753]
# Feature request

## New feature motivation
I am currently implementing automatic releases for my project, which is a web application. The project currently satisfies my needs as I can disable the npm plugins and replace it with gitlab plugins.

The problem is that I need to put the version of my project into the built application (via Webpack's DefinePlugin), but the version field of my package.json is always set to 0.0.0.

I would like to be able to use semantic-release to print the version that would be published and use it in webpack DefinePlugin.

Currently, the only way to do that would be to parse the output of a dry-run, but it is complicated and not stable.

## New feature description

I see two ways this could happen : 
 * The simple one is to provide a cli option to instruct semantic release to just print the version. For example `semantic-release --print-next-version`
 * Allow a more extensive programmatic usage of semantic-release, with a JS api that would (among other things) allow to retrieve the next version.

I could open a PR for the first solution if you'd like. The second one is a bigger task (probably several tasks)

**Comment by gr2m** [comment]
I think you can run a custom script in the `prepublishOnly` script, see https://semantic-release.gitbooks.io/semantic-release/docs/support/FAQ.html#how-can-i-use-a-npm-build-script-that-requires-the-packagejsons-version- 

At this point the version in package.json is set. Your custom script can read it and write it into your build files

**Comment by BenoitAverty** [comment]
I would but unfortunately I don't publish with npm. My build script is executed way before the semantic release.

**Comment by pvdlg** [comment]
You can also use the [@semantic-release/exec](https://github.com/semantic-release/exec) to run a script at any step you'd like that can be passed the new version number. Something like that:
```json
{
  "release": {
    "prepare": [
      "@semantic-release/npm",
      {
        "path": "@semantic-release/exec",
        "cmd": "my-script ${nextRelease.version}",
      },
    ],
  }
}
```

With this config the npm plugin is going to update your `package.json` first, then `my-script` will be called with the new version as first argument.

**Comment by BenoitAverty** [comment]
The problem with all these solutions is that I need to have two configurations : One that outputs the version or sets it somewhere, and one for the actual release in gitlab (tag, etc.) which doesn't seem to be possible easily.

For now I think i'll parse the output of a dry-run, but this is really brittle if the output changes.

**Comment by pvdlg** [comment]
> The problem with all these solutions is that I need to have two configurations : One that outputs the version or sets it somewhere, and one for the actual release in gitlab (tag, etc.) which doesn't seem to be possible easily.

Do you mean two semantic-release config? Can you clarify why would you need two configs?

**Comment by BenoitAverty** [comment]
I will try to be more detailed about my workflow.

I have a Gitlab-CI pipeline that does this jobs in order : 
```
1) Build my application (npm run build && docker build)
2) Deploy a review environment in kubernetes
3) Run e2e tests against this review environment
4) If the tests are OK && the branch is master, release with semantic-release 
```

The problem is that the customer wants to see the version number inside the application (and maybe the changelog later). Wich means I need the version number during step 1, but I don't want to release yet (the tests could still fail).

The two semantic-release configurations would allow me to use `@semantic-release/exec`
```
Config one : only exec during prepare that does npm version ${nextRelease.version}
Config two : the real configuration that does changelog, version bump, gitlab release, etc.
```

But the preferred method would just be to have a command that outputs the next version. I would use that during step one.

I hope it's more clear, don't hesitate to ask for more details.

**Comment by pvdlg** [comment]
Thanks for the clarification.

semantic-release is build around the following ideas:
- A release consist in running test, defining the version then releasing
- A given release happens only once

It seems your steps 1 and 2 are actually tests, but handled like a release. In semantic-release, it has to be either a test step or a release step. If it's a test, there is no version determined, if it's a release there is one.

If you need the version number for step 1 and 2 then it's a release. Your workflow could be viewed as doing two different release:
- One "pre-release" or "release candidate" that will go through integration testing
- One final release if the integration test are successful

We usually handle this case via npm dist-tag. Until #563 get implemented the support of dist-tag is fairly simple unfortunately as we support one dist-tag config in the `package.json` and pre-releases are not supported.

Once #563 is implemented you could do the following workflow:
- Set up the branches `master` and `rc`
- Configure `master` to release on the `@latest` dist-tag
- Configure `rc` to release on the `@rc` dist-tag
- Configure semantic-release to do the whole release (build, version bump, changelog, gitlab release, deploy on kubernetes, run e2e tests, etc...) 
- Push your new code (i.e. merge your PRs) into the `rc` branch
- Once a release released on `@rc` passes all the integration tests, merge the `rc` branch into `master`, semantic-release will make the `@rc` release available on `@latest`

This way:
- Every time you run your CI and there is new relevant commits you make a release
- Each release candidate is actually a real release (with git tag, gitlab release, changelog etc...)
- All your releases go through the integration test process
- Only the satisfying releases are made available on `@latest` and used by the final users

With the current implementation you can do something that involve some manual steps:
- Configure the dist-tag `@rc` in your `package.json`'s `publishConfig.tag` property
- Configure semantic-release to do the whole release (build, version bump, changelog, gitlab release, deploy on kubernetes, run e2e tests, etc...)
- Push your new code (i.e. merge your PRs) into the `master` branch as your currently do
- When a `@rc` release passes the test and is considered ready to be made available to users, make it available on `@latest` with the `npm dist-tag add` command

**Comment by BenoitAverty** [comment]
I can't use this workflow unfortunately. My project is not a NPM library and it's not released on any NPM registry. My real artifact is a docker image.

Moreover, the entire release process can't be executed in one go because of infrastructure constraints: some steps (deploying the review app for example) must be executed in a particular context to be able to deploy to kubernetes, the deploy part must be in a single gitlab job so I can re-deploy or rollback with gitlab's features, etc.

I understand this may not really be the use case this library is intending to solve, so if you think it's outside the scope, I would understand and you can feel free to close this issue (I'll be able to hack something and solve my problem).

I still think it could be useful to have some way to configure the script's output to make it easier to parse when doing a dry run (Doing a dry run is the way to solve my use case as mentionned in the faq by the way).

for example 
```
$ semantic-release -d --quiet --show-version
next version: 0.2.0
```
 or something like that (my example is ugly, only to illustrate)

**Comment by pvdlg** [comment]
In most cases where such feature was requested there was a better solution to implement the desired workflow, so it hasn't been implemented so far.

It seems the workflow explained here is very specific and uncommon. However we might consider adding such feature if there is other workflow that would require it.

So if anyone using a workflow that would benefit from that feature, please add a comment to this issue to explain your workflow with as much detail as possible.

**Comment by jhabdas** [comment]
> In most cases where such feature was requested there was a better solution to implement the desired workflow, so it hasn't been implemented so far.

I'm not sure I completely agree with that statement. In previous versions this was generally hacked together using `pre` and `post` commands. I know because I spent a few hours last year cross-linking a number of related issues in the repo.

I'm upgrading right now (taking about 10 new majors, whatever) and don't have much time to look at the doc. I'll try out @BenoitAverty's workaround as having the next version available is imperative for me to properly update the minified license header in my rollup library.

**Comment by jhabdas** [comment]
I tried @BenoitAverty's workaround and ran into #722. I'm not planning on digging in any further but I have to admit I was a bit startled to see `semantic-release` requiring write perms to my git repo to do a dry run as it in my mind completely negates the purpose of doing a dry run in the first place.

Nevertheless we all have options and my hat is off to the project maintainers for continuing to move this product forward under their own vision.

**Comment by jhabdas** [comment]
> In most cases where such feature was requested there was a better solution to implement the desired workflow, so it hasn't been implemented so far.

I'm not sure I completely agree with that statement. In previous versions this was sometimes hacked together using `pre` and `post` sub commands, e.g.

```
"release": "semantic-release pre && npm run build && npm publish && semantic-release post"
```

Obviously not an ideal solution. So I'm happy to see there's FAQ now and some other approaches as mentioned in the first few responses above.

@BenoitAverty imho if you're not using this product for NPM please consider looking for a less opinionated product, many of which have been mentioned in older issues related to version numbers.

**Comment by gr2m** [comment]
I agree with @pvdlg, we can leave the issue open for some more time to see if others have a similar workflow, but I can’t recall anyone form the past

**Comment by BenoitAverty** [comment]
@jhabdas I tried other libs but semantic-release is the only one that is this high-level. I thought it would be less work to implement the compte workflow (parse commits, determine version bump, generate changelog, release) by configuring semantic-release that by redoing the same work with, for example, conventional-changelog.

I still think it was easier, but maybe it's less elegant because it's a lot of configuration.

**Comment by jhabdas** [comment]
Thanks for the info, @BenoitAverty. Hope to hear back if you find a good setup with GitLab as I understand they have a number of CI/CD tools. Given you're running a Docker set-up I'm sure you have a lot of interesting hurdles to get the system set-up in a fault-tolerant way.

**Comment by andresmijares** [comment]
the feature would be extremely useful for me as well. I need the next version printed in several places and run scripts with this value.

**Comment by pvdlg** [comment]
@andresmijares could you explain your workflow and why such feature would be useful to you?

**Comment by andresmijares** [comment]
Sure thing, I'm using the package to manage release for another project in a different language, this is going to build the proyect and publish in our artifact repository name as the version, which mean that in order to build the distribution I need to pass the version number. I've tried to create a script that takes this version number and make something like:

```javascript
"release": {
   "prepare": [{
       "path": "@semantic-release/exec",
       "cmd": "./build.sh --release -v ${version}"
   }]
}
```

which I thought it would work but I havent had any luck with it.

**Comment by pvdlg** [comment]
Can you clarify what "I havent had any luck with it" means?
Is it not working? What's the problem?

**Comment by pvdlg** [comment]
btw, as indicated in [semantic-release/exec#usage](https://github.com/semantic-release/exec#usage) the proper syntax is `${nextRelease.version}` not `${version}`.
So your config should be:
```js
"release": {
   "prepare": [{
       "path": "@semantic-release/exec",
       "cmd": "./build.sh --release -v ${nextRelease.version}"
   }]
}
```

**Comment by andresmijares** [comment]
my bad, I just figured I had 1 issue not related to the package, it worked like charm, I decided to use `${version}` since it is the one marked for `tag-format` sample.

Thank you for the comment!

**Comment by deyceg** [comment]
@BenoitAverty I do something like you, and end up just using a bit of sed/awk magic to cat the next version (if there is one) to a `VERSION` file.

We have an npm library, that is consumed by some app, which is deployed to a review environment if on a feature branch. This allows us to publish the npm package under package_name@next when developing against a feature

**Comment by brad-jones** [comment]
Similar to @BenoitAverty I also work with a similar gitlab workflow, where by a release is first deployed to an environment for automated & manual testing before it proceeds on to get promoted to a production release and deployed.

Right now the solution is 100% custom, in this pipeline there is a job called `calc-next-version` which runs well before a potential release is made.

I have been using semantic release for some of my other side projects. It works well in the open source space, where generally you are not deploying anything but just releasing a new version to the public.

To be useful in a commercial / internal environment this feature would be helpful.

**Comment by pvdlg** [comment]
To implement this workflow I'd recommend to look at #991.

You could work on a `rc` branch and configure it as a pre-release branch. This way it would be release as `x.y.x-rc.n` and you can test that version. Once the test pass you can merge the `rc` branch into `master` and release the regular version.

Here is an example workflow that includes what you try to achieve: https://github.com/semantic-release/semantic-release/blob/beta/docs/recipes/pre-releases.md

**Comment by simlu** [comment]
I would very much like to see this feature as well if that is still on the table: `semantic-release --print-next-version`

I can probably hack around it, but it would make my life a lot easier!

**Comment by travi** [comment]
@simlu could you expand on some details of your use case? i'm sure it would be helpful to understand what it is that you are trying to accomplish that other plugins, especially [exec](https://github.com/semantic-release/exec), can't already help you with.

**Comment by simlu** [comment]
@travi Oh I'm sure I can do it using exec. However that will require me to juggle with multiple `.releaserc.json` files. The use case it: On dev branch do a release as circleci artifact with the release number as eg `1.2.3-rc1`, `1.2.3-rc2`, etc and on master branch do a release as github artifact as `1.2.3`.

**Comment by pvdlg** [comment]
@simlu as indicated in the issue https://github.com/semantic-release/github/issues/152 that you open this is supported in v16.0.0-beta.
You have an example similar in this doc: https://github.com/semantic-release/semantic-release/blob/beta/docs/recipes/pre-releases.md

**Comment by simlu** [comment]
@pvdlg Uhm what? How is that now related? Sorry I don't understand how pre release is related to this conversation...

**Comment by pvdlg** [comment]
You mentioned in your previous comment that you want to release version like `1.2.3-rc1`. In semantic versioning this format is called a pre-release: https://semver.org/#spec-item-9.

**Comment by simlu** [comment]
@pvdlg Ah, I see where the confusion comes from then. I can not use semantic release to release as circle ci artifacts since that can't be done as a terminal command. It's configured in the circle ci config directly. So the rc postfix was just an example.

Hence unfortunately that will not solve my use case.

**Comment by simlu** [comment]
@travi I've put a PR up here: https://github.com/semantic-release/semantic-release/pull/1076

It seems the `skipTag` is necessary unfortunately, unless I'm missing something.

**Comment by simlu** [comment]
There is a "hacky" workaround here. Basically create a folder e.g. `release` with a `.releaserc.json` and the content

```
{
  "plugins": ["@semantic-release/commit-analyzer"],
  "verifyConditions": false,
  "prepare": [{
    "path": "@semantic-release/exec",
    "cmd": "echo -n \"${nextRelease.version}\" > version.txt && exit 1"
  }],
  "npmPublish": false,
  "publish": false,
  "fail": false,
  "success": false
}
```

Then run `(cd release && semantic-release --branch=$BRANCH_NAME) || true` where `$BRANCH_NAME` is the environment variable containing your branch name.

This will work on any branch. The downside is that you don't get immediately notified / debug information when / if something goes wrong. So you might want to check for the presence of `version.txt` after.

**Comment by pvdlg** [comment]
Again, if what you want to do is to releases like `1.2.3-rc1`, `1.2.3-rc2` from `dev` and releases like `1.2.3` from `master` the workflow you are looking for is https://github.com/semantic-release/semantic-release/blob/beta/docs/recipes/pre-releases.md.

Outputting the version without publishing a release is at best a convoluted and brittle way to achieve the implementation you describe. Unless you actually want to do something else...But it's hard to say as the only explanation you provided is https://github.com/semantic-release/semantic-release/issues/753#issuecomment-455273645.

As requested in https://github.com/semantic-release/semantic-release/issues/753#issuecomment-455241656 please explain what is your objective (as in your overall goal for managing releases, not the implementation you try to make) otherwise your requests will continue to be ignored.

**Comment by simlu** [comment]
@pvdlg As [described](https://github.com/semantic-release/semantic-release/issues/753#issuecomment-455666575) it comes down to the fact that I can not use semantic-release to [store CircleCi artifacts](https://circleci.com/docs/2.0/artifacts/), since this is not done via a terminal command.

The requirements are:
1. I want to build an artifact that internally includes the release version, so when it's run by a user it's immediately clear (1) what release version this was going towards, i.e. it's clear when something is outdated and (2) what CircleCi build number is associated, i.e. it's easy to debug where the artifact came from exactly
2. Store the artifact in CircleCi
3. Use a version number that includes the CircleCi build number as well as the upcoming release number for above reasons.

Please let me know if you have any further questions about this use case!

I'm ok with the "hack" that I have in place. It is a hack, but should continue to work going forward unless something fundamentally changes with plugin handling.

**Comment by jhabdas** [comment]
@simlu please consider moving your problem info a new issue. thanks

**Comment by thekiwi** [comment]
I'm having a similar challenge with semantic release, where the output is also a docker image. @pvdlg you mentioned in another thread that semantic release was very much a 'release' tool not a 'build' tool but in our case, we need the version number at build time to package into our docker file, so that the application is able to report version information with exceptions to Sentry at runtime. 

Here's the config we're currently looking at:
```JavaScript
{
  branch: 'master',
  plugins: [
    '@semantic-release/commit-analyzer',
    '@semantic-release/release-notes-generator',
    ['@semantic-release/npm', { npmPublish: false }], // set package.json version
    ['@semantic-release/exec', {
      verifyConditionsCmd: 'docker login -u $DOCKER_USER -p $DOCKER_PASS',
      prepareCmd: 'docker build -t myorg/$CIRCLE_PROJECT_REPONAME:${nextRelease.version} .',
      publishCmd: 'docker push myorg/$CIRCLE_PROJECT_REPONAME'
    }],
    '@semantic-release/github'
  ]
}
```

As you can see, we need to do the 'determine which version' step to update our package, then run our docker build (which is multi-stage and includes things such as unit tests) and then the output of the docker build is published before finally publishing release notes to GitHub. The docker build may also fail some tests, which will currently appear as 'release failed'

Being able to run semantic-release just to determine the next version (or run it in a two-step mode) would allow us to decouple our build pipeline from the semantic-release config/process. 

Having said that, if there is a better/recommended way to handle this use case, I'm all ears :)

**Comment by simlu** [comment]
@thekiwi We still use our approach that I've outlined above and it works great. Should work well for your use case as well. Just uncollapse some replies!

**Comment by thekiwi** [comment]
@simlu Thanks! We did take a look at your solution, however as we're planning to roll this out to a couple hundred repos I was hoping for a more official solution from the Semantic Release team. Also, an earlier comment asked for people to add their use cases for this functionality to the thread, so I thought I'd add ours.

**Comment by FixRM** [comment]
Hi! Same problem here. My project is not NPM package and I need to only know current version number. All other steps like tagging, testing, packaging etc. already automated with CI. I'm just looking for way to calculate version based on commit history.

p.s. May be someone can advise other tool if this one doesn't support this?

**Comment by travi** [comment]
@FixRM is there a reason that the exec pluging, mentioned above, couldnt accomplish your goal?

**Comment by FixRM** [comment]
Hello @travi, may be I'm doing something wrong as I'm not familiar with the tool, but as far as I understand plugins are only hooks for the pipeline. I tried to add the following to package.json:

```json
"release": {
    "repositoryUrl": "http://localhost",
    "plugins": [
      [
        "@semantic-release/exec",
        {
          "prepareCmd": "echo ${nextRelease.version}"
        }
      ]
    ]
```
And then run `npx semantic-release`. But It ask me for repository Uri and accesses and it looks like it still going to tag and push something. I don't need all that stuff, it is done by CI system. All I want - is to get version number.

Am I doing something wrong? Is there a way to skip all that steps and execute only required one?

**Comment by AndyOGo** [comment]
The reason we need it is for coverage reports, which we want to tag with the next version and send it to [sonarqube by setting `sonar.projectVersion` within `sonar-project.properties` file](https://docs.sonarqube.org/latest/analysis/scan/sonarscanner/#header-2) even before we have a working release.

May I misunderstood something in the docs 🤔 
But from what I see in the [docs only the **verifyConditions** and **prepare** step](https://semantic-release.gitbook.io/semantic-release/usage/plugins#plugins-implement-actions-for-release-steps) is responsible for pushing to a remote repo:

> Responsible for verifying conditions necessary to proceed with the release: configuration is correct, **authentication token are valid**, etc...

> Responsible for preparing the release, for example creating or updating files such as package.json, CHANGELOG.md, documentation or compiled assets and **pushing a commit**.

If that is correct, I would consider it a bug, if these checks are always executed outside of any related plugin, even still if those steps are disabled by disabling CLI's plugin options, like:
```sh
npx semantic-release --verify-conditions false --verify-release false --generate-notes false --prepare false --publish false --analyze-commits @semantic-release/commit-analyzer --dry-run
```

unning `npx semantic-release --verify-conditions false --verify-release false --generate-notes false --prepare false --publish false --analyze-commits @semantic-release/commit-analyzer --dry-run` still verifies git permission and throws **EGITNOPERMISSION** even though this should be part of the **verify-conditions** plugin.

This is currently hard coded in:
https://github.com/semantic-release/semantic-release/blob/1eb302545e3ed16044cb0533430d39ce0eeed9e2/index.js#L62-L78

CLI output:
```sh
npx semantic-release --verify-conditions false --verify-release false --generate-notes false --prepare false --publish false --analyze-commits @semantic-release/commit-analyzer --dry-run
npx: installed 736 in 12.398s
[12:39:47 PM] [semantic-release] › ℹ  Running semantic-release version 15.13.21
[12:39:47 PM] [semantic-release] › ✔  Loaded plugin "analyzeCommits" from "@semantic-release/commit-analyzer"
[12:39:47 PM] [semantic-release] › ⚠  Run automated release from branch master in dry-run mode
[12:39:50 PM] [semantic-release] › ✖  The command "git push --dry-run https://github.com/axa-ch/ui-render-standalone.git HEAD:master" failed with the error message remote: Invalid username or password.
fatal: Authentication failed for 'https://github.com/axa-ch/ui-render-standalone.git/'
.
[12:39:50 PM] [semantic-release] › ✖  EGITNOPERMISSION The push permission to the Git repository is required.
semantic-release cannot push the version tag to the branch master on remote Git repository with URL https://github.com/axa-ch/ui-render-standalone.git.

Please refer to the authentication configuration documentation to configure the Git credentials on your CI environment and make sure the repositoryUrl is configured with a valid Git URL.

{ SemanticReleaseError: The push permission to the Git repository is required.
    at module.exports (/Users/axawinterthur/.npm/_npx/3537/lib/node_modules/semantic-release/lib/get-error.js:6:10)
    at run (/Users/axawinterthur/.npm/_npx/3537/lib/node_modules/semantic-release/index.js:77:11)
    at process._tickCallback (internal/process/next_tick.js:68:7)
  name: 'SemanticReleaseError',
  code: 'EGITNOPERMISSION',
  details:
   '**semantic-release** cannot push the version tag to the branch `master` on remote Git repository with URL `https://github.com/axa-ch/ui-render-standalone.git`.\n\nPlease refer to the [authentication configuration documentation](https://github.com/semantic-release/semantic-release/blob/master/docs/usage/ci-configuration.md#authentication) to configure the Git credentials on your CI environment and make sure the [repositoryUrl](https://github.com/semantic-release/semantic-release/blob/master/docs/usage/configuration.md#repositoryurl) is configured with a [valid Git URL](https://git-scm.com/book/en/v2/Git-on-the-Server-The-Protocols).'
```

**Comment by pvdlg** [comment]
> The reason we need it is for coverage reports, which we want to tag with the next version and send it to sonarqube by setting sonar.projectVersion within sonar-project.properties file even before we have a working release.

That doesn't seems to be a good practice and something we want to support. By doing that you create a coverage report referencing a release that doesn't exist. Once this specific release is made, it might contains different code, and you end up with an erroneous coverage report.

> If that is correct, I would consider it a bug, if these checks are always executed outside of any related plugin, even still if those steps are disabled by disabling CLI's plugin options, like:

You are talking about 2 different and unrelated things.

`verifyConditions` and `verifyRelease` are plugins steps. They allow user to implement their own verifications.

Verifying the authorization on the Git repo is something done in the core because we want to do that all the time no matter what plugin is used. This is necessary to make sure we can create a tag.
semantic-release uses tags to figure what commit have been added since the last release. If we can't create tags we can determine that on the next run and semantic-release couldn't possible work.

**Comment by AndyOGo** [comment]
@pvdlg Thanks for your answer and your patience.

Regarding sonarqube, we want to report if a future release is failing or succeeding, therefore we need the next version number.

Just to get your point right.
Locally one can always create as many tags in git as someone feels right, only if someone want's to push those tags by `git push --tags` to a remote git repo, he will need to make sure that he has the proper permissions.

In our case we will never push any tag to any repo, which is the reason I disabled aforementioned release steps.

I'm sure I don't know all the details of the awesome magic this tool is doing behind the scene.
Though it seems not ideal to me if git permission are checked, even though I disabled the related steps (if I understand correctly what is written in the docs and if the docs are complete and accurate).

Kind regards, Andreas

**Comment by pvdlg** [comment]
The primary purpose of semantic-release is to make release, not to be used as part of the tool chain analyzing commits to determine version.
That's why making the release (and therefore creating and pushing the tag) is not optional.

The other reason we don't allow that is because is releases have to be immutable. A release define a version and a codebase, or in other terms once a version is associated with a codebase state it can't be change. We don't want to have version `1.2.3` with a given codebase and another `1.2.3` release somewhere else with a slightly different codebase.

That's why I said that using a version number of a release not already made is a bad practice. It opens the door to discrepancy.

In your case, producing a coverage report for a version that doesn't exists yet is wrong because nothing guarantee that the codebase used for the release will be the same that the one used for the test coverage. If you are are producing a report for an unreleased codebase, then you should't name that report after a version.

**Comment by AndyOGo** [comment]
@pvdlg Thanks for your great anwser 👍 

I agree a release has to be pure, without any side effects. That's the reason I picked your tool.

Regarding the version number you are also right. Correctly we would be tagging them as incremental release candidates, if failing and as the actual release if succeeding.

Thanks again for your patience and detailed answer.

**Comment by AndyOGo** [comment]
Another use case we have for **version** is to add a `version` property to our bundle at build phase before the actual release.

**Comment by dominykas** [comment]
> Another use case we have for version is to add a version property to our bundle at build phase before the actual release.

Same here. We need to have the version number statically inside a JS bundle which is then released as a docker container, i.e. webpack and `docker build` happen way before `semantic-release`.

I realize there are hooks that would allow us to build both - the bundle and the container as part of `semantic-release` process, however that means duplicating code in CI and having a complete separate CI configuration for PR and master builds or to force `semantic-release` to be the "build agent" and make it also produce build artifacts at PR stage, but without publishing them.

**Comment by dominykas** [comment]
One workaround is to use `semantic-release` as a library (although not sure this is officially supported):

```
const semanticRelease = require('semantic-release');

const result = await semanticRelease({ dryRun: true });

if (result) {
    console.log(result.nextRelease.version);
}
```

As an aside, I'd love to see `semantic-release` expose more of the functionality to be used as a library (e.g. an option to just load the config or the commits - it would be handy building tooling for use cases that are outside of scope for it).

**Comment by jamesovens** [comment]
We could also do with this in our workflow which we run on CircleCI, I'll give the steps we take here:
1.) A pull request is created and the title must be in conventional commit form
2.) All our tests are run and a temporary environment created for UAT
3.) Pull request is merged to master which triggers the master workflow, this deploys to other pre-prod envs and then determines if a release is needed and if so deploys to production
4.) We run the release tool before we deploy to production which does everything we want in terms updating package.json and creating github tags

The reason we want to have access to the output version number is so that I can in a step before the production release determine if we actually have anything 'releasable'.

Because the pull request could just be for updated test or documentation it will be tagged as 'docs', 'test' or 'chore' and inline with conventional commit format that means this change does not need releasing. Unfortunately in CircleCI you cannot conditionally choose at runtime whether to do some jobs or not, we have to say always run the release tool and then always do the deployment.

This leads us to the scenario where the release tool will decide there are no changes and so it doesn't do anything and then the production deployment proceeds anyway which is obviously no good.

We have a way to cancel the rest of the workflow in CircleCI (the part that deploys to prod) if we can determine using some bash script if the deployment is not necessary, and we could do that if either the version number was output (and we compare it to the current version number) or we had some output that tells us that a release is not needed.

**Comment by jamesovens** [comment]
Oh and since I wrote the above I realised that I also need the version number available before I build the app because the app I am working on exposes the version number to users (same requirements as other commenters have posted).

If anyone knows of a different tool that could provide this functionality I would be happy to use that instead. I know conventional-changelog provides this, however that tool has another bug that prevents me using it (which is that it determines there is a patch version update necessary when there isn't: https://github.com/conventional-changelog/conventional-changelog/issues/399)

**Comment by gr2m** [comment]
> I also need the version number available before I build the app because the app I am working on exposes the version number to users 

I've built a plugin for that and use it in many of my projects: https://github.com/gr2m/semantic-release-plugin-update-version-in-files

You can see it in action e.g. here: https://github.com/octokit/core.js

**Comment by jamesovens** [comment]
Thanks for replying, I took a look at your plugin, it seems I cannot find a way to get your plugin to run without also creating the tags at the same time. If I tell semantic release to run in dryRun mode then that also skips the prepare step of your plugin, if I don't run in dryRun mode then your plugin runs (I think) but it also creates the git tags.

Thanks for letting me know about your plugin 👍 

Incidentally for anyone else looking for a solution to this particular problem, @simlu provided a solution that works pretty well IMO above, it was marked as off topic but here is the link to it:
https://github.com/semantic-release/semantic-release/issues/753#issuecomment-455806195

**Comment by martdavidson** [comment]
Here's my use case. We use logging software in our prerelease branches (qa, staging, etc) so that we can easily recreate a tester's session.

We need to know what version of our software the QA tester was using. This is impossible at the moment as semantic-release is run _after_ our app is built, but we need to know what the next tag is going to be in that prerelease branch to provide that to our build step.

**Comment by prvdph** [comment]
@pvdlg I don't wan't to create yet another issue regarding the git authorization being hardcoded in the core, so I'll comment on this one instead.

> Verifying the authorization on the Git repo is something done in the core because we want to do that all the time no matter what plugin is used. This is necessary to make sure we can create a tag.

And what if using a plugin is actually **the only way** to create a tag?
I have a paranoid devops, and no push access to the repository from the CI environment - it can only read. 
The CI runner has also access to some custom made api with set of strictly limited actions that can be performed on the repository itself, and amongst those actions is the tag creation.
So you can see, I'd like to create the tag via plugin to keep the semantic-versioning working properly, but I can't use it at all due to enforced write access authorization / git push verification.

**Comment by xesf** [comment]
I had a similar need of skip tagging when not using the git plugin. (not sure this is a bug though, not using git plugin, the tag still happens).

So I end-up having a prepare command that will execute the command I need to make the additional changes and use the git plugin to commit all changed files.

I manage to have instead of using semantic-release in the beginning of the CI process, and push it to the end.

Just in case someone needs my current setup, these are the plugins I used:
`plugins: [
        '@semantic-release/commit-analyzer',
        '@semantic-release/release-notes-generator',
        '@semantic-release/changelog',
        ['@semantic-release/npm', { npmPublish: false }],
        [
            '@semantic-release/exec',
            {
                prepareCmd: 'yarn release-bumpapps', // update iOS and Android versions
            },
        ],
        [
            '@semantic-release/git',
            {
                assets: [
                    'android/app/build.gradle',
                    'ios/TechTools/Info.plist',
                    'ios/TechToolsTests/Info.plist',
                    'package.json',
                    'CHANGELOG.md',
                ],
                message: 'chore(release): ${nextRelease.version} [skip ci]\n\n${nextRelease.notes}',
            },
        ],
        '@semantic-release/github',
    ],`

**Comment by GopherJ** [comment]
I cannot really image why it's so hard to implement such a simple feature, so many people want it including me. It makes semantic-release much more flexible.

**Comment by JoA-MoS** [comment]
Use Cases (any language / framework that requires the build number at build time) 
* C#
* Java / Gradle (gradle.properties)
* There are probably a lot more
* Bundling your app into a docker image - As a requirement our services are required to expose an application info endpoint that displays the application version and some other basic information. To do this we write a json file with the properties including version number to the dist directory. This is done by a custom semantic-release plugin. Unfortunately the build of the docker image happens before semantic release is run as building a releasable item should happen outside of semantic release. So to solve this we implemented a sharable config (--extends) with just the plugins needed to analyze commits and our custom plugin. This works great, except that it tags the commit. 

I thought I had found something with this https://github.com/semantic-release/semantic-release/pull/1076 but it didn't get merged. 

Looks like I will need to add this to my forked version

**Comment by Gyomu** [comment]
I'd say there are two things that made me consider the need for this feature:
1. how come everything related to git is in the git plugin, but tagging is handled by the core ? It would make much more sense to apply the single responsibility principle.
2. I stumbled on this use case recently: semantic-release will handle the release when I create a release branch (using trunk based development flow) ; but if semantic-release analyze the commits after the release branch is created and the relevant release pipeline is triggered, how can I determine this branch name ?

To me the 2nd point is what should in itself motivate such a feature: if I want to automate my release and the versioning of said release, I need to also automate the release branch creation step with correct naming (e.g. release/1.2.x ). Otherwise I would be mixing manual intervention (creating a release branch 1.2.x) that could be different than what the semantic would require (but detecting a commit triggering a 2.0.x release instead).

Hence, calling twice the semantic-release process would be best:
1. determine next version but don't tag (because we are on master): then use your CI tool to create the actual release branch
2. on the release branch, do the whole semantic-release process, from analyze to actual release/publish.

**Comment by Gyomu** [comment]
While waiting for a cleaner way to achieve previously mentioned use case, here is a one liner writing the next version in a file:
```
npx -p @semantic-release/exec -p semantic-release semantic-release --plugins "@semantic-release/commit-analyzer,@semantic-release/exec" --analyzeCommits @semantic-release/commit-analyzer --verifyRelease @semantic-release/exec --verifyReleaseCmd 'echo ${nextRelease.version} > nextRelease.txt'
```

Hope this help some of the others' needs.

**Comment by kyranjamie** [comment]
Also finding myself wanting this feature.

Without it, a developer has to ask the question "When I merge this PR into a release branch, what version will be published?". That's slow and error-prone. 

"Oh shit, I meant to remove that `BREAKING CHANGE`, it isn't actually breaking". Too late, it's published.

This feature gives developers much more confidence to publish, and make the whole process of releasing more predictable/transparent.

**Comment by Clumsy-Coder** [comment]
I'd like for a semantic-release flag to be used to ONLY output the next release version 

For now I have to jump around hoops to make it work.

`yarn semantic-release -d | grep 'Release note' | grep -Po "(\d+)\.(\d+)\.(\d+)"`

plugins
- @semantic-release/commit-analyzer
- @semantic-release/release-notes-generator

Better to have a command like `semantic-release --nextVersion` to display the next release version

**Comment by patoncrispy** [comment]
> While waiting for a cleaner way to achieve previously mentioned use case, here is a one liner writing the next version in a file:
> 
> ```
> npx -p @semantic-release/exec -p semantic-release semantic-release --dry-run --plugins "@semantic-release/commit-analyzer,@semantic-release/exec" --analyzeCommits @semantic-release/commit-analyzer --verifyRelease @semantic-release/exec --verifyReleaseCmd 'echo ${nextRelease.version} > nextRelease.txt'
> ```
> 
> Hope this help some of the others' needs.

@Gyomu that looks like exactly what I need but when I ran the command I got a 'Bad Substitution' issue which I presume relates to the `verifyReleaseCmd` parameter. I'm not great at bash and have tried escaping various characters but no luck. Any ideas there?

**Comment by Gyomu** [comment]
> > While waiting for a cleaner way to achieve previously mentioned use case, here is a one liner writing the next version in a file:
> > ```
> > npx -p @semantic-release/exec -p semantic-release semantic-release --dry-run --plugins "@semantic-release/commit-analyzer,@semantic-release/exec" --analyzeCommits @semantic-release/commit-analyzer --verifyRelease @semantic-release/exec --verifyReleaseCmd 'echo ${nextRelease.version} > nextRelease.txt'
> > ```
> > 
> > 
> > Hope this help some of the others' needs.
> 
> @Gyomu that looks like exactly what I need but when I ran the command I got a 'Bad Substitution' issue which I presume relates to the `verifyReleaseCmd` parameter. I'm not great at bash and have tried escaping various characters but no luck. Any ideas there?

Hello, normally if it's launched inside a bash environment, it should be ok.
In the end as I'm building an automated process from start to finish I did not go with the one liner solution which was more like a proof of concept before diving too deeply with semantic-release if it wasn't fitting my needs.

I configured the whole thing using a `.releaserc.json` file, especially with the `@semantic-release/exec` plugin part calling separate and dedicated scripts:
```json
["@semantic-release/exec",
  {
    "analyzeCommitsCmd": "./.release/analyzeCommitsStep.sh ${lastRelease.version}",
    "verifyReleaseCmd": "./.release/verifyReleaseStep.sh",
    "prepareCmd": "./.release/prepareStep.sh ${nextRelease.version}",
    "publishCmd": "./.release/publishStep.sh ${nextRelease.version}"
  }
],
```
Where you can see I'm using the fact that `$lastRelease` and `$nextRelease` will be provided by semantic-release itself.
I delegated the responsibility of checking what I want/need based to real scripts that could be a bit more complex that the one liner I proposed.

Actually I prefer this approach because you can verify and validate your scripts independently of the triggering process. Semantic-release is the tool analysing commits and providing with information such as the next version number, while the part where I use that is separate and can follow it's own process of modification/validation.
Then depending on your script's return code and the rest of the configuration, semantic-release will continue or fail the release process eventually.

**Comment by karlderkaefer** [comment]
according to [FAQ](https://semantic-release.gitbook.io/semantic-release/support/faq) you can simply run with `--dry-run`, the next version will be printed out
```
npx semantic-release --dry-run
>> [3:03:16 PM] [semantic-release] › ℹ  The next release version is 1.1.4
```
then you can go ahead and parse the output and reuse the next version and pass it to pipeline steps
```
export NEXT_VERSION=$(npx semantic-release --dry-run | grep 'The next release version is' | sed -E 's/.* ([[:digit:].]+)$/\1/')
echo $NEXT_VERSION

>> 1.1.4
```

**Comment by drtobbyas** [comment]
> Similar to @BenoitAverty I also work with a similar gitlab workflow, where by a release is first deployed to an environment for automated & manual testing before it proceeds on to get promoted to a production release and deployed.
> 
> Right now the solution is 100% custom, in this pipeline there is a job called `calc-next-version` which runs well before a potential release is made.
> 
> I have been using semantic release for some of my other side projects. It works well in the open source space, where generally you are not deploying anything but just releasing a new version to the public.
> 
> To be useful in a commercial / internal environment this feature would be helpful.

Absolutely, semantic-release is fine for open source libraries where only releases are important. However, in commercial applications, deployment is necessary. Automatic and manual test is employed to ensure software quality before hitting production, It makes sense to have an easy way to print next version and use that for tracking across various places. 
My workflow involves building docker images tagged with the next version number before embarking on manual testing and then generate changelog and other artifacts ready to push to production.

**Comment by JoA-MoS** [comment]
If you are interested in switching to semantic-release-plus a fork of semantic release, the functionality has been added to run semantic release but skip the git tagging.

https://semantic-release-plus.gitbook.io/semantic-release-plus/recipes/utility/expected-next-version

https://semantic-release-plus.gitbook.io/semantic-release-plus/usage/configuration#skiptag

**Comment by Blackclaws** [comment]
I also stumbled across semantic-release because I really like the idea of having version numbers detached from human emotions. However in our case we are building a Unity application and need to build both an android and an ios binary out of it. 

So our complete process is something like:
1. Determine Version of Build
2. Set version number in unity
3. Perform Unity Build
4. Perform Android and iOS builds
5. Semantic Release Tag
6. Deploy build artifacts to app stores

Unfortunately this cannot all be done on the same machine let alone in one process. Especially building iOS apps requires a macOS machine while the Unity Build takes place in a Kubernetes cluster. The build number also needs to be set before the build process starts.

I totally understand that the initial use case for semantic-release is basically releasing an npm package and the language of semantic-release is identical to the package system being used. There are a lot of cases where this luxury doesn't exist.

Having to hack around instead of just having a simple option that prints the expected version is a bit sad. Even if semantic-release is opinionated it appears to be in widespread enough use that you might consider all the people that don't use it for npm packages but have build systems that semantic-release simply doesn't support or have a multistage pipeline in their CI system that requires the version before a final release.

**Comment by srosato** [comment]
I also have a workflow use case using **next.js** and the **nx** monorepo tool.

In next.js it is easy to add an app version at build time using the `generateBuildId` function in `next.config.js`.

For example:

```js
module.exports = withNx({
  generateBuildId: async () => {
    const pkgVersion = JSON.parse(readFileSync(`${__dirname}/../package.json`).toString()).version
    return semver.parse(pkgVersion).version
  },
  webpack: (config, { buildId }) => {
      config.plugins.push(
      new webpack.DefinePlugin({
        'process.env.APP_VERSION': buildId
      })
    );

    return config;
  },
});
```

And then, in say, a gitlab pipeline file, I usually would run the following:

```
yarn nx affected:lint --base=$AFFECTED_BASE --parallel
yarn nx format:check --base=$AFFECTED_BASE --parallel
yarn nx affected:test --base=$AFFECTED_BASE --parallel

if [ $CI_COMMIT_REF_NAME == "$CI_DEFAULT_BRANCH" ]; then
    yarn semantic-release
fi

yarn nx affected --target=export --base=$AFFECTED_BASE --parallel
```

The problem is, the build can still fail at exporting and generating the static files. But then I will have already added the tag to the repository. What I would like to do is write to `package.json`, generate the `CHANGELOG.md`, then export and make my actual release with next.js and if it passes, then we can publish the tag. If it fails at exporting, nobody will care because the version will never be released so it eliminates the concern about mutable releases.

**Comment by JoA-MoS** [comment]
@srosato you can check out my approach here

https://semantic-release-plus.gitbook.io/semantic-release-plus/recipes/utility/expected-next-version

You will have to use the semantic-release-plus fork of semantic-release - https://www.npmjs.com/package/semantic-release-plus

This may also interest you - https://semantic-release-plus.gitbook.io/semantic-release-plus/recipes/monorepos/nx-monorepo

**Comment by gabisrm** [comment]
Hi,
I've read over the comments and I think I'm ahaving a similar issue. Our project is Gitlab based, where the pipeline is:

1. Build the code with the next version calculated by semantic release. This version needs to be included in package.json, as it is used across the application.
2. Tests are executed
3. If successful, then it is deployed in a dev environment for our QA department to test
4. When QA gives the OK, we publish to the production environment. At this point, the version may be rejected, and a new version must be published, restarting the workflow again.
5. On this stage, a Gitlab Release and tag is created and pushed.

What I really need is to execute the plugins independently, on different stages:

- On stage 1, execute @semantic-release/commit-analyzer and @semantic-release/npm with npmPublish: false
- On stage 5, execute   @semantic-release/release-notes-generator and @semantic-release/gitlab to publish to gitlab.

Is there any chance that semantic-release allows for independent execution of the configured plugins?

**Comment by Blackclaws** [comment]
Yes it does (kinda). Basically use --dry-run.

We have the following step:
```yaml
semantic-versioning:
  stage: determine-version
  image: node:latest
  script:
    - npm i
    - npx semantic-release --dry-run
    - test -f version.env
  artifacts:
    reports:
      dotenv: version.env
```

With this .releaserc.yaml
```yaml
plugins:
   - "@semantic-release/commit-analyzer"
   - "@semantic-release/release-notes-generator"
   - [ "@semantic-release/changelog",
      {
         "changelogFile": "CHANGELOG.md"
      } ]
   - - "@semantic-release/exec"
     - verifyReleaseCmd: "./.release-utils/verifyRelease.sh ${nextRelease.version}"
   - - "@semantic-release/gitlab"
     - gitlabUrl: https://redacted
   - - "@semantic-release/git"
     - assets: [ "CHANGELOG.md", "package.json" ]
       message: "chore(release): ${nextRelease.version} [skip ci]\n\n${nextRelease.notes}"
```

and this script:
```bash
echo "NEXT_VERSION=$1" > version.env
```

I have tried to get it to work with directly passing an echo command to semantic release but that didn't work due to shell issues.

**Comment by kristof-mattei** [comment]
@Blackclaws This doesn't work in a PR.

I did some more research and stumbled upon https://github.com/release-it/release-it which has that out of the box. 

I'm gonna do some testing but will probably switch over, this toolkit seems a lot more flexible than `semantic-release`.

**Comment by DanySK** [comment]
I hit this issue as well. It surprises me that it is so requested and that it was integrated in a fork, but it can't make it into the main distribution.

**Comment by kristof-mattei** [comment]
@DanySK which fork?

**Comment by Blackclaws** [comment]
> @DanySK which fork?

https://semantic-release-plus.gitbook.io/semantic-release-plus

**Comment by JoA-MoS** [comment]
To be specific this recipe https://github.com/semantic-release-plus/semantic-release/blob/master/docs/recipes/expected-next-version.md

**Comment by maxiride** [comment]
Chiming in the conversation to bring up another way to achieve the result described in the subject.

This solution uses Github workflow specific commands, namely the [example-defining-outputs-for-a-job](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-defining-outputs-for-a-job).


1. Use the exec plugin to store the semantic release `${nextRelease.version}` variable in a github workflow step output:
In .releaserc.json;
```json
...
    ["@semantic-release/exec",
      {
        "publishCmd": "echo ::set-output name=nextVer::${nextRelease.version}"
      }
    ]
```
2. Declare a new **joboutput**  in the workflow yml file:
```yml
...
      - name: run semantic release
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        run: npx semantic-release
        id: ver
    outputs:
      output1: ${{ steps.ver.outputs.nextVer }}
```

3. Use the step output in another job:
```yml
  print:
    needs: release
    runs-on: ubuntu-20.04
    steps:
      - name: print version
        run: echo ${{ needs.release.outputs.output1}}
```

---

# Full running example
**Workflow:** https://github.com/maxiride/semantic-release-test/blob/4d330f2d46408f2019c33661f7f67f3e836f7bd1/.github/workflows/outputs.yml

**Semantic Release configuration file:** https://github.com/maxiride/semantic-release-test/blob/4d330f2d46408f2019c33661f7f67f3e836f7bd1/.releaserc.json

**Example run:** https://github.com/maxiride/semantic-release-test/actions/runs/1755300704

---

This method is pretty hacky but gets the job done. Nonetheless a semantic release built-in method would be more future proof to use. An official Github Action with proper inputs and outputs would be even more appreciated.

**Comment by DanySK** [comment]
This is a very nice strategy too, the difference with the force-push approach are that:
1. those cloning the repo do not get the annotated tag
2. processes requiring an annotated tag to work (e.g., based on  `git describe` won't work)

**Comment by dgroh** [comment]
Much easier like suggested earlier:

```
export NEXT_VERSION=$(npx semantic-release --dry-run | grep 'The next release version is' | sed -E 's/.* ([[:digit:].]+)$/\1/')
echo $NEXT_VERSION
```

Still, waiting for a proper solution!

**Comment by Twipped** [comment]
@maxiride Have you worked out a way to do this in a pull request? I keep running into a problem where it does not recognize the PR as being one of the approved branches, even if I do `branches: ['*']`

**Comment by nunocalaim** [comment]
thanks for everyone that showed how to get the version parsed, or created a fork, or created a one liner. This was very helpful. Thanks also for the original developers of the tool! You did an amazing job! It would just be more amazing if you incorporated this massive feedback of adding this simple functionality

**Comment by paambaati** [comment]
> An official Github Action with proper inputs and outputs would be even more appreciated.

Not official, but an action that makes the next version available as a step output on Github Actions is here –
https://github.com/felipecrs/semantic-release-export-data

**Comment by doughlass** [comment]
An issue that is 5 years old and we still don't have an official way of outputting the value.

**Comment by n0th1ng-else** [comment]
Here is a simple 5-line workaround to write your own semantic-release plugin. for example 

```javascript
// <project-root>/release-exec.js
import { writeFileSync } from "fs";

const verifyRelease = async (_, context) => {
  const next = context.nextRelease.version;
  // do whatever you want with this new version. let's say we save it into .VERSION file
  writeFileSync(".VERSION", next);
};

export default {
  verifyRelease,
};
```

And then in the semantic-release config file

```javascript
// <project-root>/release.config.js
export default {
  plugins: [
    "@semantic-release/commit-analyzer",
    "@semantic-release/release-notes-generator",
    "@semantic-release/github",
    "./release-exec.js",
  ],
};
```

**Comment by ManuelRauber** [comment]
Adding my vote here as well.

It's a pretty common use case to show the version number and possibly the change log in the app, that must be created before the build.

**Comment by vinmassaro** [comment]
I am using this plugin to update the version number in a file in the repo: https://github.com/jpoehnelt/semantic-release-replace-plugin

**Comment by haarism** [comment]
> It seems the workflow explained here is very specific and uncommon. However we might consider adding such feature if there is other workflow that would require it.

Scenario: 
Nx running builds against a github action matrix for a bunch of apps/libraries/packages. 
After each build, saving artifact and downloading them in another job which publishes the artifact takes time.  You will see that in gitHub actions often this is true based on size of files you want to save. So instead each build-X publishes the build artifact to an external artifactory (or say S3) , to publish this it need the future version number.  So I agree that at this stage the version no is invalid, but let the consumer of semantic version decide when to dry run and when to run the final ?

If one of the app/lib fails, we don't want to bump the mono-repo version, and it is easy to remove/invalidate the artifact file.
Or we have to do a version '-pending' and renaming the artifact when all builds succeed. 
However, in my case some of these are apps and want to  start using these builds in pre-prod layer and a tentative version number is needed for that. Once version becomes invalid, we can roll back the build deployed- or label it dirty. All this is for speeding up the release pipelines.

So if along with --dry --no-ci which we are grateful for, can we get a `--json` option ?

**Comment by kristof-mattei** [comment]
> In most cases where such feature was requested there was a better solution to implement the desired workflow, so it hasn't been implemented so far.
> 
> It seems the workflow explained here is very specific and uncommon. However we might consider adding such feature if there is other workflow that would require it.
> 
> So if anyone using a workflow that would benefit from that feature, please add a comment to this issue to explain your workflow with as much detail as possible.

For me, what's important is that I ship artifacts and not source code. If a certain build is approved by QA, vetted by security etc, then it is THAT build that should be released.

A build of the same source code is not the same. Because of all the systems that we work with reproducible builds aren't a thing yet. 

We tag the artifact with all the checksums of the tools used, and they might differ if we choose to release a 3 day old build. Rebuilding them just for the sake of releasing introduces unnecessary risk.

So when we build something, we pre-calculate the version number and add it in the binary. 

When we then choose to release a certain tag we do not need to rebuild the binary. It is tagged with a commit and a version number. 

And we get to tell our customers that the binary they're getting is the same the one we tested throughout all our environments.

**Comment by haarism** [comment]
@kristof-mattei  I have the following workflow based on “trunk based development” on a monorepo that has many web apps and libs. 

Workflow 
1. quick merges to main with “as good as we can do” build which pass QA automation and DevSecOps pipeline
2. There are several merges in parallel and merge group and merge queue of enterprise github is used to merge to main 
3. Main is considered prod ready although not released to PROD. Any feature merged either works at prod scale or fails and blocks teams who are constantly on top of integegating things fast as priority. So all main merge is deployed to PREPROD continuously.
4. While that deployment is propogated (20 min in amplify a static generation representative pages and rest incremental) further manual and deep scans need to be done on the artifact - this is a release candidate this need to have a semetic version. And if this final deep QA and Secuity fails the build is reverted from PREPROD. Integrate fast and fail fast is the motto.

This is not a core engine or some complex application - its mostly a bunch of high performing static websites that rely on shared libraries. Here I also don't want to just semantic upgrade only for release/*app branches. I need to be able to leverage main as a source of tracking.

I'm open to feedback if you think I can use semantic release any other way. 

Currently I'm scraping semantic release dry run output. A machine readable  json seems to be lesswrong than having none.

**Comment by zachspar** [comment]
You can use [this plugin](https://www.npmjs.com/package/semantic-release-gha-output) that sets GitHub Action output variables to be consumed in later steps or jobs.

**Comment by P0pzi** [comment]
Clearly a feature that's being asked for, but is being ignored, seemingly due to arrogance.

I'm also docker-izing an application and could do with abit of shell to spit the next version out into a file, so my application can display the version to the user.

But instead I'm spending an hour googling and working out how to come up with a custom script to spit out what I need prior to the docker build to ingest it, so if we can avoid doing that, then why aren't we? This feature is a few lines of code.

I'll be using [Semantic Releases Plus](https://github.com/semantic-release-plus/semantic-release-plus) until this is added.

Flawed logic aside, thank you for the otherwise fantastic little tool.

**Comment by michaelchambaud-eaton** [comment]
After hours investigating and searching for a solution to this, I made something relatively simple.

`.releaserc` 

```
{
    "branches": ["main"],
    "plugins": [
        "@semantic-release/commit-analyzer",
        "@semantic-release/release-notes-generator",
        "@semantic-release/github",
        [
            "@semantic-release/exec", {
                "verifyReleaseCmd": "echo \"version=${nextRelease.version}\" >> $GITHUB_OUTPUT"
            }
         ]
    ]
}
```

in your workflow

```
jobs:
  get-version:
    runs-on: ubuntu-latest

    outputs:
      version: ${{ steps.semantic-release.outputs.version }}

    steps:
      - name: Checkout
        uses: actions/checkout@v4
        with:
          fetch-depth: 0

      - name: Setup Node
        uses: actions/setup-node@v4
        with:
          node-version: latest

      - name: Install `@semantic-release/exec` plugin
        run: npm install @semantic-release/exec

      - name: Semantic Release dry-run
        id: semantic-release
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        run: npx semantic-release --dry-run

      - name: Version
        run: echo "The next version is ${{ steps.semantic-release.outputs.version }}"
```

**Comment by zachspar** [comment]
> My use case is that I need to build multiple assets and include the version prior to generating a release. 

@michaelchambaud-eaton, you could use [this plugin](https://github.com/zachspar/semantic-release-gha-output) I made that outputs the version and tag. Works with dry-run too!

It uses GitHub's actions JS library to output the variables within a workflow.

**Comment by kristof-mattei** [comment]
> My use case is that I need to build multiple assets and include the version prior to generating a release.

Same here. Since I'm dealing with compiled code I can't go in and change something in my asset after release. It violates provenance.
